lib/querystring.js

```diff
@@ -310,142 +329,145 @@ function parse(qs, sep, eq, options) {
+            if (--pairs === 0)
+              return obj;
+            lastPos = i + 1;
+            sepIdx = eqIdx = 0;
+            continue;
+          }
+        } else {
+          if (lastPos < end) {
+            value += qs.slice(lastPos, end);
+            if (valEncoded)
+              value = decodeStr(value, decode);
           }
-        } else if (i === 1) {
-          // A pair with repeated sep could be added into obj in the first loop
-          // and it should be deleted
-          delete obj[key];
+          if (keyEncoded)
+            key = decodeStr(key, decode);
+        }
+
+        // Use a key array lookup instead of using hasOwnProperty(), which is
+        // slower
+        if (keys.indexOf(key) === -1) {
+          obj[key] = value;
+          keys[keys.length] = key;
+        } else {
+          const curValue = obj[key];
+          // A simple Array-specific property check is enough here to
+          // distinguish from a string value and is faster and still safe
+          // since we are generating all of the values being assigned.
+          if (curValue.pop)
+            curValue[curValue.length] = value;
+          else
+            obj[key] = [curValue, value];
         }
         if (--pairs === 0)
-          break;
+          return obj;
         keyEncoded = valEncoded = customDecode;
-        encodeCheck = 0;
         key = value = '';
-        posIdx = lastPos;
+        encodeCheck = 0;
         lastPos = i + 1;
         sepIdx = eqIdx = 0;
       }
-      continue;
     } else {
       sepIdx = 0;
-      if (!valEncoded) {
-        // Try to match an (valid) encoded byte (once) to minimize unnecessary
-        // calls to string decoding functions
-        if (code === 37/*%*/) {
-          encodeCheck = 1;
-        } else if (encodeCheck > 0 &&
-                   ((code >= 48/*0*/ && code <= 57/*9*/) ||
-                    (code >= 65/*A*/ && code <= 70/*F*/) ||
-                    (code >= 97/*a*/ && code <= 102/*f*/))) {
-          if (++encodeCheck === 3)
-            valEncoded = true;
+      // Try matching key/value separator (e.g. '=') if we haven't already
+      if (eqIdx < eqLen) {
+        if (code === eqCodes[eqIdx]) {
+          if (++eqIdx === eqLen) {
+            // Key/value separator match!
+            const end = i - eqIdx + 1;
+            if (lastPos < end)
+              key += qs.slice(lastPos, end);
+            encodeCheck = 0;
+            lastPos = i + 1;
+          }
+          continue;
         } else {
-          encodeCheck = 0;
+          eqIdx = 0;
+          if (!keyEncoded) {
+            // Try to match an (valid) encoded byte once to minimize unnecessary
+            // calls to string decoding functions
+            if (code === 37/*%*/) {
+              encodeCheck = 1;
+              continue;
+            } else if (encodeCheck > 0) {
+              // eslint-disable-next-line no-extra-boolean-cast
```

Comment:
If anyone is curious, this was added because there seems to be a positive difference when explicitly converting to a boolean in the conditional vs. implicitly converting the numeric value (or `undefined`) to a boolean.